internal/auth/providers/okta.go

```diff
@@ -0,0 +1,416 @@
+// - compares allowed groups to users' groups.
+// Conditionally returns an empty (nil) slice or a slice of the matching groups.
+func (p *OktaProvider) ValidateGroupMembership(email string, allowedGroups []string, accessToken string) ([]string, error) {
+	if accessToken == "" {
+		return nil, ErrBadRequest
+	}
+
+	if len(allowedGroups) == 0 {
+		return []string{}, nil
+	}
+
+	userinfo, err := p.UserInfo(accessToken)
+	if err != nil {
+		return nil, err
+	}
+	if len(userinfo.Groups) == 0 {
+		return nil, fmt.Errorf("no group membership returned for %s", email)
+	}
+
+	matchingGroups := []string{}
+	for _, x := range allowedGroups {
+		for _, y := range userinfo.Groups {
+			if x == y {
+				matchingGroups = append(matchingGroups, x)
+			}
+		}
+	}
+	return matchingGroups, nil
+}
+
+// UserInfo takes in an access token and sends a request to the /userinfo endpoint.
+// Conditionally returns nil response or a struct of specified claims.
+func (p *OktaProvider) UserInfo(AccessToken string) (*UserInfoResponse, error) {
+	// https://developer.okta.com/docs/api/resources/oidc/#userinfo
+	response := &UserInfoResponse{}
+	if AccessToken == "" {
+		return nil, ErrBadRequest
+	}
+	var bearer = "Bearer " + AccessToken
+	header := http.Header{}
+	header.Set("Authorization", bearer)
+
+	err := p.oktaRequest("GET", p.UserInfoURL.String(), nil, []string{"action", "userinfo"}, header, &response)
+	if err != nil {
+		return nil, err
+	}
+
+	return response, nil
+}
+
+// RefreshSessionIfNeeded takes in a SessionState and
+// returns false if the session is not refreshed and true if it is.
+func (p *OktaProvider) RefreshSessionIfNeeded(s *sessions.SessionState) (bool, error) {
+	if s == nil || !s.RefreshPeriodExpired() || s.RefreshToken == "" {
+		return false, nil
+	}
+
+	newToken, duration, err := p.RefreshAccessToken(s.RefreshToken)
+	if err != nil {
+		return false, err
+	}
+	logger := log.NewLogEntry()
+
+	s.AccessToken = newToken
+
+	s.RefreshDeadline = time.Now().Add(duration).Truncate(time.Second)
+	logger.WithUser(s.Email).WithRefreshDeadline(s.RefreshDeadline).Info("refreshed access token")
+
+	return true, nil
+}
+
+// RefreshAccessToken takes in a refresh token and returns the new access token along with an expiration date.
+func (p *OktaProvider) RefreshAccessToken(refreshToken string) (token string, expires time.Duration, err error) {
+	// https://developer.okta.com/docs/api/resources/oidc/#token
+
+	params := url.Values{}
+	params.Add("client_id", p.ClientID)
+	params.Add("client_secret", p.ClientSecret)
+	params.Add("refresh_token", refreshToken)
+	params.Add("grant_type", "refresh_token")
+
+	var response struct {
+		AccessToken string `json:"access_token"`
+		ExpiresIn   int64  `json:"expires_in"`
+	}
+
+	err = p.oktaRequest("POST", p.RedeemURL.String(), params, []string{"action", "refresh"}, nil, &response)
+	if err != nil {
+		return
+	}
+
+	token = response.AccessToken
+	expires = time.Duration(response.ExpiresIn) * time.Second
```

Comment:
let's make sure this time math is correct

Reply:
As far as I can tell, this looks correct! But let me know if there's anything in particular you're concerned about.